### PR TITLE
Handle parenthesis expansion and simplification in C# interpolations

### PR DESCRIPTION
**Pull request by DustinCampbell**
In string interpolations, parentheses can't be removed if they surround
some expression that includes a : or :: token. Othewise, a parsing
ambuity will be introduced due to the : that delimits the format clause
(i.e. $"{number:x4}").

Fixes Issue #724.

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs [hunk 1]
```diff
@@ -122,19 +123,61 @@ public static bool CanRemoveParentheses(this ParenthesizedExpressionSyntax node)
                 : false;
         }
 
+        [ThreadStatic]
+        private static Stack<SyntaxNode> s_nodeStack;
```
Comment (Pilchie) [inline src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: Why does this need to be a field - I don't see it's state shared across calls to `RemovalMayIntroduceInterpolationAmbiguity`.    Can't it just be a local?

Comment (DustinCampbell) [inline src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: This is potentially called for every single parenthesized expression. Seems a bad idea to allocate a stack each time.

Comment (Pilchie) [inline src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: Then an ObjectPool<T> is our standard approach.

Comment (DustinCampbell) [inline src/Workspaces/CSharp/Portable/Extensions/ParenthesizedExpressionSyntaxExtensions.cs]: K -- fixed

**Comment by Pilchie** [comment]
Thanks!  I get worried when I see `ThreadStatic` in a world of `Tasks` because there is no guarantee that continuations will run on the same thread.  I know this case was safe, but I'd rather just use our established patterns.

LGTM!

**Comment by basoundr** [comment]
:+1:

**Comment by mattwar** [comment]
:+1:

**Pull request merged**